docs/performance/selects.txt

```diff
@@ -0,0 +1,90 @@
+==================
+Select Performance
+==================
+
+
+Aggregations and Group By
+=========================
+
+A common use case for CrateDB is to do ``GROUP BY`` queries on tables for
+analytic purposes. Say you're selecting the ``avg``, ``max`` and ``min`` of
+some measurements over a billion records and group them by a device id.
+
+By default CrateDB processes all matching records, which may require a lot of
+processing power depending on the data set and the size of the CrateDB cluster.
+
+Similar products sometimes limit the amount of records that are processed for
+GROUP BY operations. That means the user might get inaccurate results, but for
+analytics this is often an acceptable trade off.
+
+To emulate this down sampling behavior, users can filter on the ``_docid``
+system column using a modulo operation.
+
+Given our previous use case, the query without any filters might look like
+this::
+
+   cr> SELECT
+      device_id,
+      max(value),
+      avg(value),
+      min(value)
+   FROM
+      measures
+   GROUP BY
+      device_id
+   ORDER BY
+      1 DESC;
+   +-----------+------------+-------------------+------------+
+   | device_id | max(value) |        avg(value) | min(value) |
+   +-----------+------------+-------------------+------------+
+   |         4 |      10000 | 5003.748816285036 |          0 |
+   |         3 |      10000 | 5005.297395430482 |          0 |
+   |         2 |      10000 | 5002.940588080021 |          0 |
+   |         1 |      10000 | 5002.216030711031 |          0 |
+   +-----------+------------+-------------------+------------+
+
+
+Now, to speed up this query we can append a ``WHERE _docid % 10 = 0`` filter::
+
+   cr> SELECT
+      device_id,
+      max(value),
+      avg(value),
+      min(value)
+   FROM
+      measures
+   WHERE
+      _docid % 10 = 0
+   GROUP BY
+      device_id
+   ORDER BY
+      1 DESC;
+
+   +-----------+------------+--------------------+------------+
+   | device_id | max(value) |         avg(value) | min(value) |
+   +-----------+------------+--------------------+------------+
+   |         4 |      10000 | 5013.052623224065  |          1 |
+   |         3 |      10000 | 4999.1253575025175 |          0 |
+   |         2 |      10000 | 5001.400379047543  |          0 |
+   |         1 |      10000 | 5035.220951927276  |          0 |
+   +-----------+------------+--------------------+------------+
+
+You'll notice that the result changed slightly, but is still fairly close to
+the original result.
+
+Note that the ``% 10`` in this example was arbitrary. The higher the number,
+the fewer records will match the query and the less accurate the result will
+be, trading accuracy for more performance.
+
+
```

Comment:
Should we include some output like the following, so that users get a better intuition about the modulo?

```
cr> select _docid, value, _docid % 2 from measures limit 5;
+--------+-------+--------------+
| _docid | value | (_docid % 2) |
+--------+-------+--------------+
|      0 |  7745 |            0 |
|      1 |  4934 |            1 |
|      2 |  8315 |            0 |
|      3 |  9957 |            1 |
|      4 |  6380 |            0 |
+--------+-------+--------------+
```

Reply:
it can't hurt!